src/tests/test_pcurves.cpp

```diff
@@ -120,10 +120,28 @@ class Pcurve_Ecdsa_Sign_Tests final : public Text_Based_Test {
             throw Botan::Internal_Error("During ECDSA signature generated zero r/s");
          }
 
-         const auto r_bytes = r.serialize();
-         const auto s_bytes = s.serialize();
+         return std::pair{r, s};
+      }
+
+      static bool ecdsa_verify(const Botan::PCurve::PrimeOrderCurve& curve,
+                               const Botan::PCurve::PrimeOrderCurve::AffinePoint& pk,
+                               const std::span<const uint8_t> msg,
+                               const Botan::PCurve::PrimeOrderCurve::Scalar& r,
+                               const Botan::PCurve::PrimeOrderCurve::Scalar& s) {
+         if(r.is_zero() || s.is_zero()) {
+            return false;
+         }
+
+         const auto z = curve.scalar_from_bits_with_trunc(msg);
 
-         return Botan::concat(r_bytes, s_bytes);
+         const auto s_inv = curve.scalar_invert(s);
+         const auto u_1 = curve.scalar_mul(z, s_inv);
+         const auto u_2 = curve.scalar_mul(r, s_inv);
+
+         const auto table = curve.mul2_setup(curve.generator(), pk);
+         const auto x = curve.mul2_vartime(*table, u_1, u_2).to_affine();
```

Comment:
r is not the x coordinate but rather the x coord reduced modulo the group order. Since this applies also to most ECC signature schemes there is a helper for this.

Reply:
d'accord. Done.